library/pk_wrap.c

```diff
@@ -486,6 +486,13 @@ static int rsa_alt_check_pair( const mbedtls_pk_context *pub, const void *prv )
     unsigned char hash[32];
     size_t sig_len = 0;
     int ret;
+    const mbedtls_pk_context* prv_context = prv;
+
+    if( prv_context->pk_info->type == MBEDTLS_PK_RSA_ALT )
```

Comment:
Minor: we don't need that check here, this function is only called if this condition is true.

Reply:
If you mean the check in pk.c - I've removed it. If you mean that currently, because of the dynamic dispatch, it will be called - yes, but this should be checked as a prerequisite - the function is callable with other arguments too. Wasn't moving the check here the point of earlier discussion?

Reply:
I think we should either keep things as they were before, or entirely remove the check for type equality here and move it to each `xxx_check_pair()` function. I think anything is the middle is less statisfying. As discussed elsewhere (long PR discussion pages are a mess on github), I think removing the type check entirely here and moving it to the invidividual functions is slightly cleaner.

Reply:
I agree with @mpg on this point - I think the type checks should be done in `pk_wrap.c`, while the `check_pair` wrapper code in `pk.c` should be generic.

Reply:
If I understand correctly, this would mean changing the interface so that check_pair functions could take the whole mbedtls_pk_context instead of the underlying  key pair context for the private key - so that the pk_info can be checked there. Is that ok?

Reply:
It's ok to change the interface of `check_pair_func` now. It's becoming a public API with this feature, but it wasn't a public API before. The interface of `check_pair_func` is already changing in this PR: before it passed only the `void *pk_ctx` part for the public key as well.

Reply:
@hanno-arm @mpg In this discussion about when to check types, it seems that your baseline is having the individual functions check types. But this is a departure from the old behavior! The old behavior was to insist that the private and public keys must be of the same class (same `pk_info` pointer), with an exception to support a private `RSA_ALT` with a public `RSA`.

I don't understand the rationale for changing the behavior here, beyond the mandatory change that opaque private keys must be matchable with transparent public keys, which extends the previous exception for `RSA_ALT`.

Reply:
@gilles-peskine-arm: @mpg has left another comment in https://github.com/ARMmbed/mbedtls/pull/1293#discussion_r166381734 which might clarify the rationale (this is not to say we must follow it, of course -- but I still think it's the cleaner approach).

Reply:
@gilles-peskine-arm I'm not arguing for changing the behaviour. We can perfectly move the checks to the individual `xxx_check_pair()` functions while preserving observable behaviour: just have each such function start with `if( prv->pk_info->type != pub->pk_info->type ) return ( MBEDTLS_PK_TYPE_MISMATCH );` (except `rsa_alt_check_pair()`) et voilà! Old behaviour preserved while moving all type-specific code from `pk.c` to `pk_wrap.c`.

It's just a trade-off between avoiding non-generic code in `pk.c` at the cost of some code duplication in `pk_wrap.c`, it's completely orthogonal to whether we want to change observable behaviour or not.

Anyway, as Hanno said in another part of the discussion, I don't think this is important enough to delay this PR, so now that I've hopefully clarified my position, I'll just defer to you and/or Andrzej for the final call and approve (in that area) the next thing pushed as long as it's not broken (like the last commit was wrt RSA_ALT).

Reply:
Updated to reflect the behaviour described by Gilles with an update to the API.